budgetportal/webflow/views.py

```diff
@@ -164,3 +284,45 @@ class ProvInfraProjectSearchView(FacetMixin, HaystackViewSet):
         "status_order",
         "estimated_completion_date",
     ]
+
+    def list(self, request, *args, **kwargs):
+        csv_download_params = self._get_csv_query_params(request.query_params)
+        response = super().list(request, *args, **kwargs)
+        response.data["csv_download_url"] = reverse(
+            "provincial-infrastructure-project-api-csv"
+        )
+        if csv_download_params:
+            response.data["csv_download_url"] += "?{}".format(csv_download_params)
+        return response
+
+    @action(detail=False, methods=["get"])
+    def get_csv(self, request, *args, **kwargs):
+        self.serializer_class = self.csv_serializer_class
+        response = super().list(request, *args, **kwargs)
+        return self.generate_csv_response(
+            response.data["results"], filename=self._get_filename(request.query_params)
+        )
+
+    def _get_csv_query_params(self, original_query_params):
+        csv_download_params = original_query_params.copy()
+        csv_download_params.pop("fields", None)
+        csv_download_params.pop("limit", None)
+        params = ""
+        for param_name, param_value in csv_download_params.items():
+            params += "{}={}&".format(param_name, urllib.parse.quote(param_value))
+        return params[:-1]
+
+    def _get_filename(self, query_params):
+        keys_to_check = (
+            "province",
+            "department",
+            "status",
+            "primary_founding_source",
+            "q",
+        )
+        extension = "csv"
+        filename = "provincial-infrastructure-projects"
+        for key in keys_to_check:
+            if query_params.get(key):
+                filename += "-{}-{}".format(key, slugify(query_params[key]))
```

Comment:
slugify key as well please (in case someone crafts a weird url)

Reply:
we use only particular keys so do not need to slugify it.
Let me know if we want to change that and use all keys from url.

Reply:
oh I see - ok cool